feat(fms): cFMS 1.5

Benjozork opened the pull request:
Credit @professoralex13 @tracernz for the help.

## Summary of Changes

* This PR introduces support for the following FMS features
  * LNAV
    * Holding patterns
    * Turn direction constraints on non-TF legs
    * Overfly restriction support
    * ARINC424 leg types
      * AF, CA, CI, CR, CF, DF, HF, HM, HA legs
    * Turn prediction types
      * Path capture
      * Course capture
      * Direct to fix turn
      * Holding pattern entry turn   
  * ND
    * Removed flight plan loading from `localStorage`
    * Corrected active waypoint ETA display
    * Added `EfisVectors` systems with optimized transmit task queue + support for **future** display of OFFSET, SECONDARY, SECONDARY DASHED, MISSED APPROACH, ALTERNATE and EOSID flight paths.

## Known limitations

- CA leg terminations are sometimes in the wrong place, and do not adapt to V/S.
- Some path captures will be incorrectly drawn. This will not affect guidance.
- INTCPT calculation can be off on large distances.
- Course captures do not adapt to PPOS when nextLeg is active. This can cause the final path to be off to the side.

## Screenshots (if necessary)

**LSZH DEGES3W Before**

![image](https://user-images.githubusercontent.com/4503241/145327972-79b57b90-c874-4f1c-9399-e25f293e4ccc.png)

**LSZH DEGES3W After**

![image](https://user-images.githubusercontent.com/4503241/145327987-249f3b25-9a6b-45ec-9ccf-ab62a58ea562.png)

**Overfly support on TF legs**

![image](https://user-images.githubusercontent.com/4503241/145328060-7aebe5f9-4c06-4e48-87c3-4ce3d025ec84.png)

**Direct To Fix turns**

![image](https://user-images.githubusercontent.com/4503241/145328072-5c2ba622-4bde-4ab4-a974-1fdef6e674e2.png)

**Honeywell path captures**

![image](https://user-images.githubusercontent.com/4503241/145328104-eed9e7be-b2f5-48e0-8e4e-506f12f34746.png)

## References

An absolutely enormous amount of them and a year+ of conception.

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): someperson#4953

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->

PHASE 1 - General LNAV and path generation

* Make sure to test loading various SIDs and STARs and that their drawing is reasonable
  * Test departures with turn direction constraints
* Make sure to test flying various SIDs and STARs and that the aircraft flies it correctly
  * Keep in mind that path capture turns (sharp lines joining the next leg with a 45 degree angle) will not be followed exactly - this is accurate to the IRL Honeywell FMS 
  * Check if turns change shape / size while flying them (they should not)
* Make sure sequencing and guidance work correctly (no turning the wrong way, minimal overshooting / undershooting)
  * Make sure VM (MANUAL) legs do not sequence and do not show a distance or ETA on the ND but do show a bearing
  * Make sure regular legs show a distance, bearing and ETA on the NDD
* Make sure ND does not crash and that there is no unacceptable delay in path drawing 

PHASE 2 - Holding patterns, entries and exits (not available yet)

TBD

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

davidwalschots commented on src/fmgc/src/efis/EfisCommon.ts:
```diff
@@ -0,0 +1,77 @@
+            return [114.1, 113.6, 113.6];
+        }
+        return [227.7, 227.2, 227.2];
+    case Mode.PLAN:
+        if (range <= 10) {
+            return [7, 7, 7];
+        }
+        if (range <= 20) {
+            return [14, 14, 14];
+        }
+        if (range <= 40) {
+            return [28, 28, 28];
+        }
+        if (range <= 80) {
+            return [56, 56, 56];
+        }
+        if (range <= 160) {
+            return [112, 112, 112];
+        }
+        return [224, 224, 224];
+    default:
+        return [0, 0, 0];
+    }
```
Random suggestion

```suggestion
    case Mode.PLAN:
        if (range <= 10) {
            return Array(3).fill(7);
        }
        if (range <= 20) {
            return Array(3).fill(14);
        }
        if (range <= 40) {
            return Array(3).fill(28);
        }
        if (range <= 80) {
            return Array(3).fill(56);
        }
        if (range <= 160) {
            return Array(3).fill(112);
        }
        return Array(3).fill(224);
    default:
        return Array(3).fill(0);
    }
```

beheh replied:
Personally I find the original more readable.

tracernz replied:
Particularly as it’s consistent with the other modes which aren’t all the same number.

Infrecon commented:
Potential bug, DEGES3W departure from LSZH RWY 16. 

Probably not yet implemented Leg Type?

<img width="641" alt="DEGES3S" src="https://user-images.githubusercontent.com/62938529/145564225-4f9f8934-ad70-4ca0-bc68-d3887f5148f0.png">
<img width="1551" alt="DEGES3S Chart" src="https://user-images.githubusercontent.com/62938529/145564234-87a2f1fe-e627-4a12-b812-f1433a0744dd.png">

lukecologne commented on src/fmgc/src/efis/EfisVectors.ts:
```diff
@@ -0,0 +1,177 @@
+
+        const engagedLateralMode = SimVar.GetSimVarValue('L:A32NX_FMA_LATERAL_MODE', 'Number') as LateralMode;
+        const armedLateralMode = SimVar.GetSimVarValue('L:A32NX_FMA_LATERAL_ARMED', 'Enum');
+        const navArmed = (armedLateralMode >> 0) & 1;
+
+        const transmitActive = engagedLateralMode === LateralMode.NAV || navArmed;
```
Active vectors should be transmitted if any lateral autocontrol mode is active, so if either NAV or LOC is armed, or if NAV, LOC, LOC*, LAND, FLARE, ROLL OUT, RWY, GA TRK or RWY TRK is active
```suggestion
        const navArmed = (armedLateralMode >> 0) & 1;
        const navActive = engagedLateralMode === LateralMode.NAV;
        const takeoffGoaroundModeActive = engagedLateralMode === LateralMode.RWY || engagedLateralMode === LateralMode.RWY_TRACK || engagedLateralMode === LateralMode.GA_TRACK;
        const locArmed = (armedLateralMode >> 1) & 1;
        const locActive = engagedLateralMode >= LateralMode.LOC_CPT && engagedLateralMode <= LateralMode.ROLL_OUT;

        const transmitActive = navActive || takeoffGoaroundModeActive || locActive || navArmed || locArmed;
```

SkylarHutter commented:
Having issues with the ND showing lines properly. Only shows the path if I am zoomed all the way out. Maybe because the fix is so far?

https://i.imgur.com/8vLlJjp.png
https://i.imgur.com/gb421gE.png


https://i.imgur.com/vU0wFyB.png
https://i.imgur.com/TxiLLRn.png

Benjozork commented:
Yes, I mistakenly implemened an EIS1 limitation which does not seem to exist anymore.

I'll revert / change soon enough.

TVOpium commented:
Potential bug : 

DTNH RWY27 LOBNA1D Departure
<a href="https://ibb.co/zrbjfYp"><img src="https://i.ibb.co/JK20k4V/charts.jpg" alt="charts" border="0"></a>
<a href="https://ibb.co/2Ngn1Zb"><img src="https://i.ibb.co/5FBMgK7/Sans-titre.jpg" alt="Sans-titre" border="0"></a>

LFBO Arrival Initial RNAV RWY32L Transition ADIMO2
<a href="https://ibb.co/XkbGf75"><img src="https://i.ibb.co/3Y7P5pF/ch.jpg" alt="ch" border="0"></a>
<a href="https://ibb.co/PtXc11S"><img src="https://i.ibb.co/FnZ588P/ch02.jpg" alt="ch02" border="0"></a>
<a href="https://imgbb.com/"><img src="https://i.ibb.co/RSHjgyn/Sans-titre.jpg" alt="Sans-titre" border="0"></a>

FinniganH commented:
> Yes, I mistakenly implemented an EIS1 limitation which does not seem to exist anymore.
> 
> I'll revert / change soon enough.

Has this been reverted as of yet? I grabbed a new download 30 mins ago, still seeing the same issue:
![2](https://user-images.githubusercontent.com/77217017/146031326-712c387a-1d61-4e99-b837-1bc5cd220446.jpeg)
![1](https://user-images.githubusercontent.com/77217017/146031329-6aaf658a-4992-4327-a2ec-9f1d91a6cb58.jpeg)

In this shot, it WAS working on 40 a few miles ago, now it's blank:
![3](https://user-images.githubusercontent.com/77217017/146031596-77704bf2-6460-46b5-b860-2d6ac0375f46.png)

Benjozork commented:
No, I have not reverted this yet;

Amorisia commented:
Hello,

Here is my test results departing from SAWH runway 25 with USU2B SID with SUNLI transition (using Navigraph AIRAC 2112 rev 2)
Here is the chart:
![USU2B 25](https://user-images.githubusercontent.com/3002706/146047713-0c117fa8-a17d-49e3-a6a8-2cadd0bbd12b.jpg)

A screenshot showing I082 misplaced:
![before TO](https://user-images.githubusercontent.com/3002706/146048035-45de5131-f65e-474e-ae04-ec63f5773c9b.jpg)

There was also a curved leg that was not part of the flight plan, morphing as I was flying, so I decided to make a video for better analysis.

This is a private video of the departure available on youtube, with some comments about what I was doing in subtitles and in the description.
[https://www.youtube.com/watch?v=TevpO_OJaN0](url)

Hope this helps and again, good job, unless these few bugs this is another big step and amount of work here !

FinniganH commented:
Modified original flightplan. Was direct VESDO then YXC, decided to go direct YXC, and this is the result:

![image](https://user-images.githubusercontent.com/77217017/146072224-cb171383-7a42-468a-bad0-a561f6ef758b.png)

Seems like it's created an invisible line from the prior fix to YXC instead of present position direct.

Benjozork commented:
I've changed the culling algorithm to exclude vectors >= 322NM in length instead of using the start / end position. This should remove **some** invisible leg issues.

This will cause issues if you have a very long leg in your flight plan. I've added this as a condition for showing the `MAP PARTLY DISPLAYED` message. This is not Airbus' actual length limit, but is a better solution until we have a better solution

![image](https://user-images.githubusercontent.com/4503241/146146704-74adbe74-efbc-45e0-9df1-4d9c3f8b3cf9.png)

alexr4339 commented:
Name: Alex R#6663
Date: 19/12/2021
Tier: 1
Version of this PR: Downloaded PR on the same day, around 12 UTC

Description: 
I did multiple test with this PR and discovered a few bugs. I assume the static altitude waypoint bug is known. Issues with the direct to are known aswell. I had the following procedure:
Load a flight on a runway, insert arrival and departure on Init A, aswell as altitude (sometimes) and flaps aswell as vspeeds on the perf page. Then I inserted the respective departure on the F-PLN page and either in the air or on the ground I inserted an arrival too.

General bugs:
Static altitude waypoint. Altitude waypoint seem to be static and dont move. This can mess a procedure severly up (known)
Direct to issues. Messes up the sequencing, no radius between t-p and direct course to fix, nor drawing of connection between t-p and fix (known)
MCDU sometimes black after loading on runway. Note that this occured in 10 to 20 % of the times when loading the aircraft onto the runway. I did not restart MSFS and it was like loading a flight, then either restart or turn back to menu and load another flight.

Procedural bugs:
LSZR KPT1M
The sequencing and description of this procedure is perfectly accurate on the flight plan page, however the drawing was wrong and the AP tried to follow the wrong drawing
![LSZRKPT1M](https://user-images.githubusercontent.com/83085903/146683034-bea945cd-c589-40bc-a609-2fed25a63c01.PNG)
![20211219151045_1](https://user-images.githubusercontent.com/83085903/146683049-56a0f87f-f01d-4257-a8d8-05f7c9674c54.jpg)
![20211219141805_1](https://user-images.githubusercontent.com/83085903/146683057-ce1bfbdc-127f-47ed-af75-b8cbe0d31b62.jpg)
.
EDXW OLDI1N
The selected procedure took longer to draw than usual (around 10s instead of 1 or 2 usually).
Another issue, after XW100 you shall turn right direct to SLT. Due to the increasing airspeed during climbout the radius constantly updates. However after having initiated the right turn, it doesnt update anymore, which will result that you dont turn directly towards the fix anymore, as your speed increeases and your "overshoot" the procedure
![EDXWOLDI1N](https://user-images.githubusercontent.com/83085903/146683229-ea6b6507-c746-4733-ab31-790adf1cdbc6.PNG)
![20211219135906_1](https://user-images.githubusercontent.com/83085903/146683236-7d68df52-f018-4666-b899-189d1ec4472c.jpg)

LSGG BELU6N
The waypoint GVA8 was drwan only at a selected range of 10. Note: The static alt waypoint issue and how it can potentially mess a procedure up can be seen here aswell
![20211219134524_1](https://user-images.githubusercontent.com/83085903/146683270-8e07fe82-adde-48f3-94a5-344c1524ebde.jpg)
![20211219134520_1](https://user-images.githubusercontent.com/83085903/146683274-fb87086b-cf9c-455f-ac43-b378ce0cc456.jpg)

Benjozork commented:
For `LSZR KPT1M`, issue identified and am thinking of a fix. The problem is that constrained turn direction is not being taken into account.

Benjozork commented:
As the priority right now is fixing the current issues in the publicly available LNAV, we will be adding manual holding to a future PR. Holds part of procedures will still work.

alexr4339 commented:
Name: Alex R#6663
Date of test: 28/12/2021
Downloaded PR on same day, at 10 UTC
Nav Data: MSFS stock

General:
This PR has matured well and is rock solid. Especially with custom waypoints its capable of nearly everything. I wouldnt want to fly without CFMS15 again and I think due to the fact that it is a drastic improvement to CFMS1 it can be merged to exp soon.

Bugs:
Sometimes legs are not connected or drawn as a right/left turn although they should be connected straight. In one case it affected the guidance (LICR GIMEL1X after LONDA towards RCA). Its not a drastic problem as a DTO solved this issue. But the incorrect drawing is surely not beautiful and might irritate users.
![20211228110431_1](https://user-images.githubusercontent.com/83085903/147587744-c6249dcb-deca-4e58-b87e-0feef6652045.jpg)
![20211228110533_1](https://user-images.githubusercontent.com/83085903/147587750-2691d98f-e460-498a-b20f-fa5b57d2a067.jpg)
It looked normal after the discontinuity was cleared
![20211228112803_1](https://user-images.githubusercontent.com/83085903/147587800-437d111d-74b2-46f4-88f3-f5f808d4526c.jpg)
This one above was weird as I was coming from north
![20211228112959_1](https://user-images.githubusercontent.com/83085903/147587881-698a741e-8ab0-4fff-8855-c69ca3394f46.jpg)
![20211228130124_1](https://user-images.githubusercontent.com/83085903/147587885-6d68e404-12b4-47a9-9ca4-5486885e3a70.jpg)
![20211228132341_1](https://user-images.githubusercontent.com/83085903/147587888-d466d46e-91b2-42b2-afdc-87990b0852c2.jpg)
You can see that LONDA and RCA were not connected and I lost NAV guidance. A simple DTO solved it

I know holdings are not part of this PR for now, but the radius was static. I know I was too fast but even after getting back on course with 200 knots the radius was too small, perhaps due to the fact it calculated the radius with 200kts albeit my TAS was greater due to the altitude.
![20211228132634_1](https://user-images.githubusercontent.com/83085903/147588031-e7863ba2-5082-44b1-91f0-83a55fac1aa2.jpg)

![20211228133148_1](https://user-images.githubusercontent.com/83085903/147588249-62795cb5-f854-4c6c-8520-aaccff93263a.jpg)
Here it drew the turn as intended
![20211228133402_1](https://user-images.githubusercontent.com/83085903/147588271-ff40f5db-edb4-4544-841b-ed4b91c6e0b9.jpg)
After having passed RCA11 it flew staright and the right turn dissapeared ( I think the drawing of the turn dissappeared as soon as it targeted RCA11. Interesting to know is that the waypoint after RCA11 is called RCA11 too. A DTO and turning right via heading mode solved this issue.

p493z3r0 commented:
Very good improvement, however i've found an Issue with the Vebit4s Departure out of LSZH. According to the charts it should be a Left turn to cross KLO at R180 and then intercept WIL R052 to BREGO. It is drawn correctly on the ND as well but the AP commands a right turn instead of a left turn. 
<img width="493" alt="image" src="https://user-images.githubusercontent.com/5236256/148373168-56a0a75a-c0ba-4d44-afa3-84ff33eb7e36.png">

tracernz commented on src/fmgc/src/guidance/LnavConfig.ts:
```diff
@@ -37,7 +47,7 @@ export const LnavConfig = {
      * Whether to draw path debug points and print them out
      */
-    DEBUG_PREDICTED_PATH: false,
+    DEBUG_PREDICTED_PATH: true,
```
Did you intend to commit this one? :D
